Add cmake build system

mtaillefumier opened the pull request:
This pr contains a minimal version of the cmake build system. It compiles the library and generate all necessary config files for use in other cmake and pkg-config projects. It should make the library detection in tmlQCD easier.

NB: there is no changes to the original build system.   

- [x] config files for cmake and pkg-config
- [x] reformatting
- [ ] add ctests

Commit:
Add cmake build system

Commit:
[cmake] Fix cmake use_package issues

kostrzewa commented:
@kottnad @chaoos in preparation for the discussion it would be great if you could take a look at this

Commit:
Build static library by default

Commit:
Install lemon_benchmark and lemon_contents

chaoos reviewed (commented):
I can build, but not install. It fails to install lemon_contents with:

```
$ cmake -S . -B build -DCMAKE_INSTALL_PREFIX=./install_dir
...
$ cmake --build build -j
...
$ cmake --install build
-- Install configuration: ""
-- Installing: /home/roman/projects/tmp/lemon/install_dir/lib/cmake/clemon/ClemonConfig.cmake
-- Installing: /home/roman/projects/tmp/lemon/install_dir/lib/cmake/clemon/ClemonConfigVersion.cmake
-- Installing: /home/roman/projects/tmp/lemon/install_dir/lib/pkgconfig/clemon.pc
-- Installing: /home/roman/projects/tmp/lemon/install_dir/lib/liblemon.so
-- Set non-toolchain portion of runtime path of "/home/roman/projects/tmp/lemon/install_dir/lib/liblemon.so" to ""
-- Installing: /home/roman/projects/tmp/lemon/install_dir/lib/cmake/clemon/ClemonTargets.cmake
-- Installing: /home/roman/projects/tmp/lemon/install_dir/lib/cmake/clemon/ClemonTargets-noconfig.cmake
CMake Error at build/cmake_install.cmake:103 (file):
  file INSTALL cannot find "/home/roman/projects/tmp/lemon/lemon_contents":
  No such file or directory.
```

It does work if the build directory is equal to the source directory though, ie.

```
cmake -S . -DCMAKE_INSTALL_PREFIX=./install_dir
cmake --build . -j
cmake --install .
```

works!

kottnad commented:
Trying to install on e.g. juwels booster still fails for me, i.e. does not move the binaries to the separate install directory :

CMake Error at cmake_install.cmake:99 (file):                                                                                                                                                                                                  file INSTALL cannot find                                                                                                                                                                                                                     "/p/home/jusers/ottnad1/juwels/source/lemon/lemon_contents": No such file                                                                                                                                                                    or directory.

(lemon_contents and lemon_benchmark are located in the build not the source directory.)

Commit:
[cmake] CMAKE_BINARY_DIR is missing

Commit:
[cmake] rename clemon target to lemon

mtaillefumier commented on cmake/config.h.in:
```diff
@@ -0,0 +1,80 @@
+#cmakedefine PACKAGE_STRING "@PROJECT_NAME @PROJECT_VERSION@"
+
+/* Define to the one symbol short name of this package. */
+#cmakedefine PACKAGE_TARNAME "clime"
```
oh. I removed the config.h.in generation entirely. configure will still generate it but it is ignored at compilation time

Commit:
[cmake] more cleanup

- Removed config.h
- Removed header files checking
- Static library is built by default (still with -fpic)
- Renamed target to lemon::lemon
-

mtaillefumier commented:
thanks a lot.

kostrzewa commented:
Thanks a lot for this!
